### ISSUE TITLE
Improve Source & Execution Engine

### ISSUE THREAD

**Issue by araddon** [araddon/qlbridge#30]
Job/Exec engine pipeline/tasks need complete refactor.  

Revamp the Plan/Exec with the following **Goals**
* align source, and planning from dataux/qlbridge
   * get rid of `SourceTask` interface in dataux move to `Accept` visitor pattern w poly-fill
* refactor sources, exec engine to be much cleaner/better/modular code

**Tasks**
* [x] Refactor the SubQuery/From to allow folding, use just source's as input to Join:  each different `from` in query is its own source & discrete task
    * [x] The TaskSequential/Parallel don't have channel routing (in, out, setup) right
    * [x] rewrite where, source, etc to support fold
      * [x] `Where` implemented in each SubSelect
      * [x] bug in sourceproject/sqldrivermessage doesn't match columns expected in query
      * [x] extract joinkey from Merge into `JoinKey` function/Task
        * [x] implement `JoinExpr` to create key used in join expression (ie, for multi-condition join conditions, create composite key similar to cql)
        * [x] create `joinNodes` during sql rewrite
        * [x] are we going to use int?  string?  for the "key" or "id"? ie ? add update to use new Key()/Id() or equivalent
      * [x] `projection` w unit test
  * [x] implement sub-queries tests depends on #35
    * [x] fix lexer which can't descend to sub-clauses
    * [x] unit test for 3 sources
    * [x] column list for merged sources (two sources feeding into 3rd)
* [x] TestSqlDbConnFailure:  lock when running two queries
* [x] Naive Parallel Planner (see below drawing) single node

![img_20150919_144144](https://cloud.githubusercontent.com/assets/7269/9978315/b23a6a36-5edc-11e5-8108-52961838c700.jpg)
